source/server/init_manager_impl.cc

```diff
@@ -38,14 +45,19 @@ void InitManagerImpl::initialize(std::function<void()> callback) {
 
 void InitManagerImpl::initializeTarget(TargetWithDescription& target) {
   TRACE_INIT_MANAGER("invoking initializeTarget {}", target.second);
-  target.first->initialize([this, &target]() -> void {
-    TRACE_INIT_MANAGER("completed initializeTarget {}", target.second);
-    ASSERT(std::find(targets_.begin(), targets_.end(), target) != targets_.end());
-    targets_.remove(target);
-    if (targets_.empty()) {
-      TRACE_INIT_MANAGER("initialized");
-      state_ = State::Initialized;
-      callback_();
+  target.first->initialize([self = self_, &target]() -> void {
+    if (*self != nullptr) {
```

Comment:
Instead of checking captured `shared_ptr<InitManager*>`, how about making `callback_` managed by [`CallbackManager`](https://github.com/envoyproxy/envoy/blob/master/source/common/common/callback_impl.h), and `remove` in destructor so the callback is never called?

Reply:
This is a little hard to talk about because there are two levels of callbacks:

* The callback passed into `InitManagerImpl`'s ctor and stored as `callback_`;
* The callbacks passed into `Target::initialize()` for each target.

The problem this is solving relates to the second one, not the first. The callbacks we pass to the targets capture `this`, and if this `InitManagerImpl` instance is deleted before one of the targets calls its callback, it's not safe to do anything with `this`.  For example, with tracing on:

```c++
TRACE_INIT_MANAGER("completed initializeTarget {}", target.second);
```

... crashes because it references `this.description_` and `target.second` (which was owned by `targets_`). And:

```c++
targets_.remove(target);
```

... crashes because `targets_` is gone. Calling the `callback_` would crash too, whether or not it was managed by a `CallbackManager`.

Reply:
Yeah I understand that. If `Target::initialize()` return a `CallbackHandle` and you call its `CallbackHandle::remove()` in the destructor, wouldn't it be same of this PR?

Reply:
Agree with @lizan that is the solution I was thinking of as well (or some variant of it). Essentially hold a handle to all pending callbacks and then if we go away, cause those callbacks to no longer be invoked. Although I agree what you have would work, I think this is a bit more idiomatic.

Reply:
Oh, I think I see what you're saying... `InitManagerImpl` would hang onto all the `CallbackHandle`s for each of its targets, and cancel them in the dtor. I'll give that a shot.

Reply:
On second thought, I don't think that approach is super clean either, at least as I understand it.

If `Target::initialize()` returned a `CallbackHandle` that the `InitManagerImpl` would hang onto for cancellation, that would mean that each target would have to be given ownership of a `CallbackManager`, rather than a raw callback. That is, you'd go from:

```c++
class Target {
public:
  virtual void initialize(std::function<void()> callback) PURE;
};
```

to:

```c++
class Target {
public:
  virtual CallbackHandle* initialize(std::unique_ptr<CallbackManager<void>> manager) PURE;
}
```

And `InitManager` would look something like:

```c++
// traces and asserts removed for clarity ...
void InitManagerImpl::initializeTarget(Target& target) {
  auto cb = [this, &target]() -> void {
    targets_.remove(target);
    if (targets_.empty()) {
      state_ = State::Initialized;
      callback_();
    }
  };
  auto cb_manager = std::make_unique<CallbackManager<void>>();
  cb_handles_.push_back(cb_manager.add(cb));
  target.initialize(std::move(cb_manager));
}

InitManagerImpl::~InitManagerImpl() {
  for (auto handle : cb_handles_) {
    handle->remove();
  }
}
```

That would work, but it seems like sort of a degenerate use of `CallbackManager`, where each target has to have its own single-use instance that contains just a single callback.

Having said that, there is definitely some overlap between `InitManagerImpl` and `CallbackManager`. There's a strong analog between `InitManagerImpl::registerTarget(Target&)` and `CallbackManager::add(Callback)` -- they are both taking some sort of deferred action and sticking it in a list. And there's some similarity between `InitManagerImpl::initialize()` and `CallbackManager::runCallbacks()` -- they are both invoking all deferred actions -- although there's some differences there. So I figure there's some design opportunity here... I'll give it some more thought.

Reply:
You don't necessarily need to use CallbackManager, but just the idea of having a handle to a callback that on destruction makes the callback go away. It may end up being easier to just do something internal to this class. (I haven't thought about it much but just throwing that out there.)

Reply:
Alright, I went a different direction, just adding a `cancel` method to `Init::Target`. Not fancy at all, easy enough to understand... Let me know what you think.

Reply:
Ok, that approach broke `ListenerManagerImplTest`, and I didn't really like it anyway. I've gone back to a `shared_ptr`-based approach, simplified a bit more this time.